service/history/api/updateactivityoptions/api.go

```diff
@@ -0,0 +1,264 @@
+
+	return &api.UpdateWorkflowAction{
+		Noop:               false,
+		CreateWorkflowTask: false,
+	}, nil
+}
+
+func applyActivityOptions(
+	mergeInto *activitypb.ActivityOptions,
+	mergeFrom *activitypb.ActivityOptions,
+	updateFields map[string]struct{},
+) error {
+
+	if _, ok := updateFields["taskQueue.name"]; ok {
+		if mergeFrom.TaskQueue == nil {
+			return serviceerror.NewInvalidArgument("TaskQueue is not provided")
+		}
+		if mergeInto.TaskQueue == nil {
+			mergeInto.TaskQueue = mergeFrom.TaskQueue
+		}
+		mergeInto.TaskQueue.Name = mergeFrom.TaskQueue.Name
+	}
+
+	if _, ok := updateFields["scheduleToCloseTimeout"]; ok {
+		mergeInto.ScheduleToCloseTimeout = mergeFrom.ScheduleToCloseTimeout
+	}
+
+	if _, ok := updateFields["scheduleToStartTimeout"]; ok {
+		mergeInto.ScheduleToStartTimeout = mergeFrom.ScheduleToStartTimeout
+	}
+
+	if _, ok := updateFields["startToCloseTimeout"]; ok {
+		mergeInto.StartToCloseTimeout = mergeFrom.StartToCloseTimeout
+	}
+
+	if _, ok := updateFields["heartbeatTimeout"]; ok {
+		mergeInto.HeartbeatTimeout = mergeFrom.HeartbeatTimeout
+	}
+
+	if mergeInto.RetryPolicy == nil {
+		mergeInto.RetryPolicy = &commonpb.RetryPolicy{}
+	}
+
+	if _, ok := updateFields["retryPolicy.initialInterval"]; ok {
+		if mergeFrom.RetryPolicy == nil {
+			return serviceerror.NewInvalidArgument("RetryPolicy is not provided")
+		}
+		mergeInto.RetryPolicy.InitialInterval = mergeFrom.RetryPolicy.InitialInterval
+	}
+
+	if _, ok := updateFields["retryPolicy.backoffCoefficient"]; ok {
+		if mergeFrom.RetryPolicy == nil {
+			return serviceerror.NewInvalidArgument("RetryPolicy is not provided")
+		}
+		mergeInto.RetryPolicy.BackoffCoefficient = mergeFrom.RetryPolicy.BackoffCoefficient
+	}
+
+	if _, ok := updateFields["retryPolicy.maximumInterval"]; ok {
+		if mergeFrom.RetryPolicy == nil {
+			return serviceerror.NewInvalidArgument("RetryPolicy is not provided")
+		}
+		mergeInto.RetryPolicy.MaximumInterval = mergeFrom.RetryPolicy.MaximumInterval
+	}
+	if _, ok := updateFields["retryPolicy.maximumAttempts"]; ok {
+		if mergeFrom.RetryPolicy == nil {
+			return serviceerror.NewInvalidArgument("RetryPolicy is not provided")
+		}
+		mergeInto.RetryPolicy.MaximumAttempts = mergeFrom.RetryPolicy.MaximumAttempts
+	}
+
+	return nil
+}
+
+func validateActivityOptions(
+	validator *api.CommandAttrValidator,
+	namespaceID string,
+	ai *persistencespb.ActivityInfo,
+	ao *activitypb.ActivityOptions,
+) error {
+	attributes := &commandpb.ScheduleActivityTaskCommandAttributes{
+		TaskQueue:              ao.TaskQueue,
+		ScheduleToCloseTimeout: ao.ScheduleToCloseTimeout,
+		ScheduleToStartTimeout: ao.ScheduleToStartTimeout,
+		StartToCloseTimeout:    ao.StartToCloseTimeout,
+		HeartbeatTimeout:       ao.HeartbeatTimeout,
+		ActivityId:             ai.ActivityId,
+		ActivityType:           ai.ActivityType,
+	}
+
+	_, err := validator.ValidateActivityScheduleAttributes(namespace.ID(namespaceID), attributes, nil)
```

Comment:
If you look closer you will see that `Validate...` not only validate but also fixes some fields in command attributes. You need to copy them back.

Reply:
good catch. I did notice that...but totally forgot.